### PR TITLE
fix(compiler): prevent underscore attr name camelcasing

### LINKED ISSUE #1369: [compiler] compiler doesn't validate public properties with underscore names
### Description
The following @api properties are legal in LWC class today:
```js
import { LightningElement, api } from 'lwc';

export default class App extends LightningElement {
    @api myProp;
    @api my_public'
}
```
Per HTML standard, the attribute names must be all lower case and may only contain hyphens, which means that in order for the component author to pass _**myProp**_ via template, the property name must become hyphenated:
```html
<c-app my-prop="foo"> <c-app/>
``` 

There is inconsistency in syntax between class property _myProp_ and its equivalent html _my-prop_, but user can still pass down the value. However, there is no way to pass down a property to match class @api with underscore in its name - **my_public**, as underscore is not a legal html attribute character.

**Actual**: User has no ability to pass down a property via template to match a component api with an underscore in its name.

**Expected**: Add compile time validation to disallow public property syntax that cannot meet HTML standard for attribute naming. In other words, prevent underscores in public properties.

### PR DESCRIPTION
**Pull request by apapko**
## Details
Fixes #1369 - prevent camel-casing of the attribute names with underscore.
Rules:
- attribute name must start with alphabetic character
- attribute name must end with alpha-numeric character
- attribute name cannot contain a combination of underscore and non-alpha-numeric characters

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

### CODE REVIEW

File: packages/@lwc/template-compiler/src/__tests__/parser.spec.ts [hunk 1]
```diff
@@ -572,6 +572,68 @@ describe('props and attributes', () => {
...
+            </template>`);
+
+            expect(root.children[0].props).toMatchObject({
+                under_scoreHyphen: { value: 'bar' },
+            });
+        });
+
+        it('attribute name separated by underscore and immediate hyphen', () => {
+            const { root } = parseTemplate(`<template>
+                <x-button under_-hyphen="bar"></x-button>
```
Comment (caridy) [inline packages/@lwc/template-compiler/src/__tests__/parser.spec.ts]: this one should probably be invalid... but hey!

File: packages/@lwc/template-compiler/src/__tests__/parser.spec.ts [hunk 2]
```diff
@@ -572,6 +572,68 @@ describe('props and attributes', () => {
...
+            </template>`);
+
+            expect(root.children[0].props).toMatchObject({
+                under_scoreSecond_underScore: { value: 'bar' },
+            });
+        });
+
+        it('attribute name with leading underscore and hyphen', () => {
+            const { root } = parseTemplate(`<template>
+                <x-button _leading-underscore="bar"></x-button>
```
Comment (caridy) [inline packages/@lwc/template-compiler/src/__tests__/parser.spec.ts]: same here, leading underscore should not be allowed in LWC. IMO

File: packages/@lwc/template-compiler/src/__tests__/parser.spec.ts [hunk 3]
```diff
@@ -572,6 +572,68 @@ describe('props and attributes', () => {
...
+            </template>`);
+
+            expect(root.children[0].props).toMatchObject({
+                _leadingUnderscore: { value: 'bar' },
+            });
+        });
+
+        it('attribute name with trailing underscore and hyphen', () => {
+            const { root } = parseTemplate(`<template>
+                <x-button trailing-underscore_="bar"></x-button>
```
Comment (caridy) [inline packages/@lwc/template-compiler/src/__tests__/parser.spec.ts]: same here... should throw...

File: packages/@lwc/template-compiler/src/parser/attribute.ts [hunk 4]
```diff
@@ -291,8 +291,11 @@ function shouldCamelCaseAttribute(element: IRElement, attrName: string) {
 export function attributeToPropertyName(element: IRElement, attrName: string): string {
     let propName = attrName;
     if (shouldCamelCaseAttribute(element, attrName)) {
-        propName = ATTRS_PROPS_TRANFORMS[propName] || propName;
+        const attrToSplit = ATTRS_PROPS_TRANFORMS[propName] || propName;
+        propName = attrToSplit
+            .split('_')
+            .map(part => camelcase(part))
```
Comment (caridy) [inline packages/@lwc/template-compiler/src/parser/attribute.ts]: this line is equivalent to: `.map(camelcase)`

File: packages/@lwc/template-compiler/src/parser/index.ts [hunk 5]
```diff
@@ -695,10 +695,31 @@ export default function parse(source: string, state: State): TemplateParseResult
...
+                    name,
+                    treeAdapter.getTagName(node),
+                ]);
+                return;
+            }
+
+            // disallow attr name with an underscore combined with non alphabetic characters
+            if (name.match(/_[^a-zA-Z]|[^a-zA-Z]_/)) {
+                const node = element.__original as parse5.AST.Default.Element;
+                warnAt(ParserDiagnostics.ATTRIBUTE_NAME_CANNOT_CONTAIN_UNDERSCORE_WITH_HYPHEN, [
```
Comment (caridy) [inline packages/@lwc/template-compiler/src/parser/index.ts]: the message name doesn't reflect the actual error... it is not longer about -_

File: packages/@lwc/template-compiler/src/parser/index.ts [hunk 6]
```diff
@@ -695,10 +695,31 @@ export default function parse(source: string, state: State): TemplateParseResult
...
+                const node = element.__original as parse5.AST.Default.Element;
+                warnAt(ParserDiagnostics.ATTRIBUTE_CANNOT_START_OR_END_WITH_UNDERSCORE, [
+                    name,
+                    treeAdapter.getTagName(node),
+                ]);
+                return;
+            }
+
+            // disallow attr name with an underscore combined with non alphabetic characters
+            if (name.match(/_[^a-zA-Z]|[^a-zA-Z]_/)) {
```
Comment (caridy) [inline packages/@lwc/template-compiler/src/parser/index.ts]: since attributes are parsed to lowercase by parse5, you only need to match `a-z` here

**Comment by caridy** [comment]
fix the tests, and let's move on.

File: packages/@lwc/template-compiler/src/parser/index.ts [hunk 7]
```diff
@@ -695,10 +695,31 @@ export default function parse(source: string, state: State): TemplateParseResult
...
+                const node = element.__original as parse5.AST.Default.Element;
+                warnAt(ParserDiagnostics.ATTRIBUTE_CANNOT_START_OR_END_WITH_UNDERSCORE, [
+                    name,
+                    treeAdapter.getTagName(node),
+                ]);
+                return;
+            }
+
+            // disallow attr name with an underscore combined with non alphabetic characters
+            if (name.match(/_[^a-z]|[^a-z]_/)) {
```
Comment (ekashida) [inline packages/@lwc/template-compiler/src/parser/index.ts]: Are we excluding numbers on purpose?

Comment (caridy) [reply packages/@lwc/template-compiler/src/parser/index.ts]: that's a good catch, it should support numbers as well... e.g.: `foo_1`

Comment (apapko) [reply packages/@lwc/template-compiler/src/parser/index.ts]: Loosened up the restriction to allow trailing numbers; but not leading.

File: packages/@lwc/errors/src/compiler/error-info/template-transform.ts [hunk 8]
```diff
@@ -464,17 +464,26 @@ export const ParserDiagnostics = {
         level: DiagnosticLevel.Warning,
         url: '',
     },
-    ATTRIBUTE_CANNOT_START_OR_END_WITH_UNDERSCORE: {
+
+    ATTRIBUTE_NAME_MUST_START_WITH_ALPHABETIC_CHARACTER: {
         code: 1124,
         message:
-            '{0} is not valid attribute for {1}. Attribute name cannot start or end with an underscore.',
+            '{0} is not valid attribute for {1}. Attribute name cannot start with non alphabetic character.',
```
Comment (caridy) [inline packages/@lwc/errors/src/compiler/error-info/template-transform.ts]: probably avoid negation on this messag,e something like: `Attribute name must start with alphabetic character.` instead

File: packages/@lwc/errors/src/compiler/error-info/template-transform.ts [hunk 9]
```diff
@@ -464,17 +464,26 @@ export const ParserDiagnostics = {
...
         level: DiagnosticLevel.Error,
         url: '',
     },
-    ATTRIBUTE_NAME_CANNOT_COMBINE_UNDERSCORE_WITH_NON_ALPHA_CHARS: {
+
+    ATTRIBUTE_NAME_MUST_END_WITH_ALPHA_NUMERIC_CHARACTER: {
         code: 1125,
         message:
-            '{0} is not valid attribute for {1}. Attribute name cannot contain combination of underscore and hyphen characters.',
+            '{0} is not valid attribute for {1}. Attribute name cannot start or end with non alpha-numeric character.',
```
Comment (caridy) [inline packages/@lwc/errors/src/compiler/error-info/template-transform.ts]: same

**Pull request merged**